Rule Builders

Jacoby6000 opened the pull request:
Allows creation of "rule builders" which I described a bit in #6. 

Essentially, you can now have structures like this:

```scala
case class InputFoo(untrimmed: String, maybeEmptyList: List[Int])
case class ValidatedFoo(trimmed: String, nonEmptyList: NonEmptyList[Int])

val rule = 
    Rule.builder[InputFoo]
      .check("untrimmed", _.untrimmed)(Rule.trimString)
      .check("maybeEmptyList", _.maybeEmptyList)(Rule.nonEmptyList)
      .build[ValidatedFoo]

rule(InputFoo("bar ", List("baz"))) // Ior.Right(ValidatedFoo("bar", NonEmptyList("baz")))
rule(InputFoo("bar", List()) // Ior.Left(NonEmptyList(ErrorMessage("Must not be empty", PField("maybeEmptyList", PNil))))
```

The important thing here, is that transformations caused by the rules are preserved.

### Motivation

Provided some user input `X`, a validated version of `X` occupies a subset of all values for `X`. If your validator is a function that ultimately returns `X`, you've lost the context about `X` which tells you it is validated.  If your validation function instead converts to a more refined type, like `Y`, then `Y` can accurately portray values for `X` that have been validated.  From there on, your code can just accept instances of `Y`, and you can rest assured that you're only dealing with validated data.  Additionally, a raw `X` may not be parameterized on it's type at all, while `Y` can be an HKT of sorts.

Commit:
Merge branch 'release/0.1.0'

Commit:
Merge branch 'release/0.1.1'

Commit:
Merge branch 'release/0.2.0'

Commit:
Merge branch 'release/0.2.1'

Commit:
Start rule-builder work

Commit:
Finish rule-builder work, add tests

Commit:
Add tests to some functions I added

Jacoby6000 commented:
@davegurnell It looks like it's been awhile since you've looked at this repo. I know the PR has only been up for a day, but I'd really like to see this (or something similar) get added. It's really the only thing stopping me from using it.

dispalt commented:
for the record, I really like this and use checklist.

davegurnell commented on checklist/src/main/scala/checklist/Rule.scala:
```diff
@@ -4,20 +4,36 @@ import cats.{Applicative, Traverse}
   def map[C](func: B => C): Rule[A, C] =
     Rule.pure(value => this(value) map func)
 
+  def emap[C](func: B => Checked[C]): Rule[A, C] =
```
I've not seen this name elsewhere. Is it a name with provenance, or is there a clearer alternative?

Jacoby6000 replied:
I stole it from [circe](https://github.com/circe/circe/blob/master/modules/core/shared/src/main/scala/io/circe/Decoder.scala#L193). I'm not sure where they got the name though.  I'm happy to call it whatever.. I'm really not sure what a good and proper name would be for this is.

davegurnell replied:
Ok. Let's merge it now and rename it on a subsequent release if it causes confusion.

davegurnell commented on checklist/src/main/scala/checklist/Rule.scala:
```diff
@@ -4,20 +4,36 @@ import cats.{Applicative, Traverse}
   def contramap[C](func: C => A): Rule[C, B] =
     Rule.pure(value => this(func(value)))
 
+  def contramapPath[C, D: PathPrefix](path: D, func: C => A): Rule[C, B] =
```
I'm wondering if this method ought to have two lists of parameters.

Jacoby6000 replied:
I'm inclined to agree. At the time of writing, all my contramap functions were so simple that I couldn't think of a reason to separate them.  I doubt that'll always be the case though.

davegurnell replied:
Can you tweak the PR for this please? Happy to merge tonight!

Jacoby6000 replied:
Done!

davegurnell reviewed (approved):
I've put a couple of questions inline. It'd be great if you could take a look at them, but otherwise this is aces.

davegurnell commented:
@Jacoby6000 sorry I missed this and your issue -- this is awesome! I'd never thought about the type refinement angle -- it's a great idea.

I've put a couple of small questions inline. Can you take a look at them? I'm more or less happy to just merge this, though.

Jacoby6000 commented:
Something I'm not entirely happy about with this is using `.check` as the method, but there are other places where `.field` is used... Both names are pretty arbitrary and have much different behavior, so I'm not really sure what to do about that. It might get confusing to end users. 

Also, in other places you use some macros to figure out the path, but here I'm not doing that.. I'm explicitly taking it in. I'd like to keep it explicit, since it makes the behavior less surprising and a bit more flexible (your function for contramap may pull multiple fields from the parent object). But I'm not sure how you feel about that, or if you have better ideas.

Jacoby6000 commented:
Oh, also I messed with your build file a bit, which is why I changed the view bounds to implicit params, and removed some old imports. I hope you don't mind that.  Just wanted to make that clear, since it touched some unrelated things.

Commit:
Change contramapPath to use two parameter lists

davegurnell commented:
Good point about the `check` method name. I didn't notice the inconsistency. I think it makes sense to keep the method names (i.e. `field`) consistent. The `builder` / `build` names you've used make sense.

I'm pretty happy with my macro approach because it's fast to compile. It doesn't give exhaustivity like the `HList` approach does, but then exhaustivity isn't always what one wants. However, I'm also a fan of choosing the right tool for the job. Shapeless makes sense for building full-class rules like in your examples. I'm happy to let the user decide what they want to do.

How about this. We organise the code so your builder/build pattern comes from a separate package with a separate import, a bit like `circe.generic.auto`. Then we can document that package separately. Users will be less confused because they're specifically opting in to using a derivation-like approach. What do you think?

I'm happy with your changes to the build file. Cross-building against Scala 2.12 is great. I may fiddle with the `scalacOptions` a bit if they cause me problems, but I doubt they will.

Jacoby6000 commented:
Sounds good to me.  Since it's going to be opt-in, I'll make it a separate sub-project, so people can pull in checklist without getting the shapeless stuff.

davegurnell commented:
Ok sounds good. Thanks very much for working on this, by the way. It's really great!

Commit:
Move shapeless-based builders/refinement to it's own sub-project and package

Commit:
Update travis scala version

Jacoby6000 commented:
No problem! Thanks for giving me something to build on.  I'm far too lazy to do the initial ground work 😉. 

Before merging, you might want to add some methods to the check stuff that use lenses. I'm not sure though.  I'm not super familiar with monocle, so I didn't use them. I'm sure they could be helpful though!

davegurnell commented:
I'll merge this now. I'm keen to tweak some method names in the future. I'll open issues to discuss and do it in a minor release, though.

davegurnell commented:
Merged and release as 0.3.0. Again, there are some minor tweaks to the syntax that I'd like to make in the future, but I don't want to get in the way of you using this stuff.

davegurnell commented:
Thanks! :+1:

Jacoby6000 commented:
@davegurnell I can see that you published 0.3.0, but it looks like the refinement stuff didn't make it to maven.  Can you take a look when you get a chance?

Jacoby6000 commented:
Oh nevermind. I was looking for `checklist-refinement` but it's in there as just `refinement`

davegurnell commented:
Ah. That's something to fix next release then :)
On Fri, 1 Sep 2017 at 19:31, Jacob Barber <notifications@github.com> wrote:

> Oh nevermind. I was looking for checklist-refinement but it's in there as
> just refinement
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/davegurnell/checklist/pull/7#issuecomment-326653135>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAOI596T6-ZnsOe9u3YxI19iOODuviXfks5seE1qgaJpZM4PFKyl>
> .
>
